generators/chipyard/src/main/scala/harness/HasHarnessInstantiators.scala

```diff
@@ -0,0 +1,96 @@
+import chipyard.{ChipTop}
+
+// -------------------------------
+// Chipyard Test Harness
+// -------------------------------
+
+case object MultiChipNChips extends Field[Int](0) // 0 means ignore MultiChipParams
+case class MultiChipParameters(chipId: Int) extends Field[Parameters]
+case object BuildTop extends Field[Parameters => LazyModule]((p: Parameters) => new ChipTop()(p))
+case object HarnessClockInstantiatorKey extends Field[() => HarnessClockInstantiator]()
+case object HarnessBinderClockFrequencyKey extends Field[Double](100.0) // MHz
+case object MultiChipIdx extends Field[Int](0)
+
+class WithMultiChip(id: Int, p: Parameters) extends Config((site, here, up) => {
+  case MultiChipParameters(`id`) => p
+  case MultiChipNChips => up(MultiChipNChips) max (id + 1)
+})
+
+class WithHomogeneousMultiChip(n: Int, p: Parameters, idStart: Int = 0) extends Config((site, here, up) => {
+  case MultiChipParameters(id) => if (id >= idStart && id < idStart + n) p else up(MultiChipParameters(id))
+  case MultiChipNChips => up(MultiChipNChips) max (idStart + n)
+})
+
+class WithHarnessBinderClockFreqMHz(freqMHz: Double) extends Config((site, here, up) => {
+  case HarnessBinderClockFrequencyKey => freqMHz
+})
+
+// A TestHarness mixing this in will
+// - use the HarnessClockInstantiator clock provide
+trait HasHarnessInstantiators {
+  implicit val p: Parameters
+  // clock/reset of the chiptop reference clock (can be different than the implicit harness clock/reset)
+  private val harnessBinderClockFreq: Double = p(HarnessBinderClockFrequencyKey)
+  def getHarnessBinderClockFreqHz: Double = harnessBinderClockFreq * 1000000
+  def getHarnessBinderClockFreqMHz: Double = harnessBinderClockFreq
+
+  // buildtopClock takes the refClockFreq, and drives the harnessbinders
+  val harnessBinderClock = Wire(Clock())
+  val harnessBinderReset = Wire(Reset())
+
+  // classes which inherit this trait should provide the below definitions
+  def referenceClockFreqMHz: Double
+  def referenceClock: Clock
+  def referenceReset: Reset
+  def success: Bool
+
+  // This can be accessed to get new clocks from the harness
+  val harnessClockInstantiator = p(HarnessClockInstantiatorKey)()
+
+  private val chipParameters = if (p(MultiChipNChips) == 0) {
+    Seq(p)
+  } else {
+    (0 until p(MultiChipNChips)).map { i => p(MultiChipParameters(i)).alterPartial {
+      case TargetDirKey => p(TargetDirKey) // hacky fix
+      case MultiChipIdx => i
+    }}
+  }
+
+  // This shold be called last to build the ChipTops
+  def instantiateChipTops(): Seq[LazyModule] = {
+    val lazyDuts = chipParameters.zipWithIndex.map { case (q,i) =>
+      LazyModule(q(BuildTop)(q)).suggestName(s"chiptop$i")
+    }
+    val duts = lazyDuts.map(l => Module(l.module))
+
+    withClockAndReset (harnessBinderClock, harnessBinderReset) {
+      lazyDuts.zipWithIndex.foreach {
+        case (d: HasIOBinders, i: Int) => ApplyHarnessBinders(this, d.lazySystem, d.portMap)(chipParameters(i))
```

Comment:
This PR doesn't include any fragments to wire up ChipTops to each other right?

Reply:
Next PR